File: Search Suggestions System/search-suggestions-system.py
class Solution(object):
    def suggestedProducts(self, products, searchWord):
        """
        :type products: List[str]
        :type searchWord: str
        :rtype: List[List[str]]
        """
        products.sort()
        res = []
        cur = ""

        for ch in searchWord:
            cur += ch
            
            left = self.binarySearch(products, cur, True)
            right = self.binarySearch(products, cur + '~', False)
            
            temp = []
            count = 0
            for i in range(left, right):
                temp.append(products[i])
                count += 1
                if count == 3:
                    break
            
            res.append(temp)
        
        return res
    
    def binarySearch(self, products, target, isLower):
        left, right = 0, len(products)
        while left < right:
            mid = left + (right - left) // 2
            if (products[mid] < target) if isLower else (products[mid] <= target):
                left = mid + 1
            else:
                right = mid
        return left

File: Search Suggestions System/search-suggestions-system.java
class Solution {
    public List<List<String>> suggestedProducts(String[] products, String searchWord) {
        Arrays.sort(products);
        List<List<String>> res = new ArrayList<>();
        StringBuilder cur = new StringBuilder();

        for (char ch : searchWord.toCharArray()) {
            cur.append(ch);

            int left = binarySearch(products, cur.toString(), true);
            int right = binarySearch(products, cur.toString() + '~', false);

            List<String> temp = new ArrayList<>();
            int count = 0;
            for (int i = left; i < right && count < 3; i++) {
                temp.add(products[i]);
                count++;
            }

            res.add(temp);
        }

        return res;
    }

    private int binarySearch(String[] products, String target, boolean isLower) {
        int left = 0, right = products.length;
        while (left < right) {
            int mid = left + (right - left) / 2;
            if (products[mid].compareTo(target) < (isLower ? 0 : 1)) {
                left = mid + 1;
            } else {
                right = mid;
            }
        }
        return left;
    }
}
